utils/wasm-gen/src/generator/syscalls/invocator.rs

```diff
@@ -173,43 +178,161 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
+            self.sys_call_imports
+                .clone()
+                .into_iter()
+                .flat_map(|(syscall, (amount, _))| {
+                    iter::repeat(syscall)
+                        .take(amount as usize)
+                        .collect::<Vec<_>>()
+                });
+
+        let insertion_mapping =
+            self.get_syscall_insertion_mapping(syscalls_to_insert, &insert_into_funcs)?;
+
+        for (insert_into_fn, syscalls) in insertion_mapping {
+            self.insert_sys_calls_into_fn(insert_into_fn, syscalls)?;
+        }
+
+        log::trace!(
+            "Random data after inserting sys-calls invoke instructions - {}",
+            self.unstructured.len()
+        );
+
+        Ok(())
+    }
+
+    /// Distributes provided syscalls among provided function ids.
+    ///
+    /// Returns mapping `func_id` <-> `syscalls which should be inserted into func_id`
+    /// as a vector to avoid indeterminance that may be caused by `HashMap` usage.
+    fn get_syscall_insertion_mapping<I>(
+        &mut self,
+        syscalls: I,
+        insert_into_funcs: &[usize],
+    ) -> Result<Vec<(usize, Vec<InvocableSysCall>)>>
+    where
+        I: Iterator<Item = InvocableSysCall>,
+    {
+        let mut insertion_mapping: HashMap<usize, Vec<InvocableSysCall>> = HashMap::new();
+        for syscall in syscalls {
+            let insert_into = *self.unstructured.choose(insert_into_funcs)?;
+
+            match insertion_mapping.entry(insert_into) {
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().push(syscall);
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(vec![syscall]);
+                }
+            }
+        }
+
+        let mut insertion_mapping: Vec<_> = insertion_mapping.into_iter().collect();
+        // Sort it to avoid indeterminance appeared because of HashMap usage.
+        insertion_mapping.sort_by(|a, b| a.0.cmp(&b.0));
+
+        Ok(insertion_mapping)
+    }
+
+    fn insert_sys_calls_into_fn(
+        &mut self,
+        insert_into_fn: usize,
+        syscalls: Vec<InvocableSysCall>,
+    ) -> Result<()> {
+        log::trace!(
+            "Random data before inserting sys-calls invoke instructions into function {insert_into_fn} - {}",
+            self.unstructured.len()
+        );
+
+        let fn_code_len = self.module.count_func_instructions(insert_into_fn);
+
+        // The end of insertion range is second-to-last index, as the last
+        // index is defined for `Instruction::End` of the function body.
+        // But if there's only one instruction in the function, then `0`
+        // index is used as an insertion point.
+        let last = if fn_code_len > 1 { fn_code_len - 2 } else { 0 };
+
+        let mut insertion_positions = syscalls
+            .into_iter()
+            .map(|syscall| Ok((self.unstructured.int_in_range(0..=last)?, syscall)))
+            .collect::<Result<Vec<_>>>()?;
```

Comment:
not sure whether more optimised, but you can collect it to min binary heap.

Reply:
do not implement if it doesn't require only changing the `collect` stuff with removing the `insertion_positions.sort_by(|a, b| b.0.cmp(&a.0));`